### PR TITLE
Fix the SDPA AOT export issue

### PR DESCRIPTION
**Pull request by sijiac**
Summary:
## Context
TL;DR: aot_export failed for SDPA memory efficient backend when using `inference_mode`

The CMF AOTI lowering started to fail on the trunk. We have the script (https://fburl.com/code/kfk64i5s) to reproduce the issue quickly (log: P1469307638). By bisecting the stack, we found the issue starting from the D58701607

## Root Cause
In the `inference_mode()`,
the `aten::scaled_dot_product_attention` was not decomposed before the `functionalization` and the op it-self was an out-place op, so the `functionalization` doesn't make change and then was decomposed into `masked_fill_.`, then decomposed to the `copy_`
So it's `aten::sdpa` --- (functionalization) ---> `aten::sdpa` --- (decompose) ---> `masked_fill_` --- (decompose) ---> `copy_` ---> failure

In the `torch.no_grad()`,
`aten::sdpa` was decomposed before `functionalization`, so the story is
`aten::sdpa` --- (decompose) ---> `masked_fill_` --- (functionalization) ---> `masked_fill` --- (decompose) ---> `out-place ops` ---> good

## How to fix
Long-term:
The issue was tracked in the ticket (https://github.com/pytorch/pytorch/issues/129418). The long-term fix could be we do one more round of `functionalization` after the `decompose`, like

`aten::sdpa` --- (functionalization) ---> `aten::sdpa` --- (decompose) ---> `masked_fill_` --- (functionalization) ---> `masked_fill` ---> good

Short-term:
It would be a big change I guess. To unblock the production use-case, I marked the `aten::sdpa` should be decomposed in this diff

Test Plan:
local repro works now

buck run mode/opt scripts/sijiac/prototypes:sdpa_aoti

Differential Revision: D59385876

### CODE REVIEW

**Review by zou3519 — APPROVED** [review]

**Comment by sijiac** [comment]
Synced with @tarun292 offline. The failure of `test_to_edge_transform_and_lower ` by this PR was unexpected. @tarun292  will do a forward fix of the test after the PR was landed

@pytorchbot merge -f "test failure was unexpected"

**Comment by huydhn** [comment]
@sijiac @albanD just points out to me that ET failure needs a forward fix to go with.  To land this safely, you could follow the steps in this wiki https://www.internalfb.com/intern/wiki/PyTorch/Using_PyTorch/Executorch/Step-by-step_Guide_to_Land_Synchronous_Changes_to_PyTorch_and_ExecuTorch/ to update both PT and ET.  The process requires  interlocking steps of updating PT and ET.

On the other hand, let me know if you want to land this and break trunk waiting for the forward fix, I could mark the job as unstable in a few hours to hide the failure in trunk.
